base/task.jl

```diff
@@ -341,31 +341,45 @@ end
 
 ## lexically-scoped waiting for multiple items
 
-function sync_end(c::Channel{Any})
-    local c_ex
-    while isready(c)
-        r = take!(c)
-        if isa(r, Task)
-            _wait(r)
-            if istaskfailed(r)
-                if !@isdefined(c_ex)
-                    c_ex = CompositeException()
-                end
-                push!(c_ex, TaskFailedException(r))
+"""
+    _sync1!(c_ex::T, waitable) -> c_ex′::T′
+
+where `T = T′ = Union{Nothing,CompositeException}`
+"""
+function _sync1!(c_ex, @nospecialize(r))
+    if isa(r, Task)
+        _wait(r)
+        if istaskfailed(r)
+            if c_ex === nothing
+                c_ex = CompositeException()
             end
-        else
-            try
-                wait(r)
-            catch e
-                if !@isdefined(c_ex)
-                    c_ex = CompositeException()
-                end
-                push!(c_ex, e)
+            push!(c_ex, TaskFailedException(r))
+        end
+    else
+        try
+            wait(r)
+        catch e
+            if c_ex === nothing
+                c_ex = CompositeException()
             end
+            push!(c_ex, e)
         end
     end
+    return c_ex
+end
+
+function sync_end(c::Channel{Any})
+    c_ex = nothing
+    while isready(c)
+        c_ex = _sync1!(c_ex, take!(c))
+    end
     close(c)
-    if @isdefined(c_ex)
+    # Racy `@spawn`/`@async` many manage to `put!` tasks in the channel `c`.
+    # Let's synchronize them as well, to report possible errors:
+    for r in c
+        c_ex = _sync1!(c_ex, r)
+    end
```

Comment:
Alternatively, maybe it's better to throw an error when we find something is scheduled/`@sync_add`ed after `close`?

Reply:
I wonder if we could somehow cheaply prevent the use of the channel from any task that is not (transitively) part of the `@sync` group. Probably not though (perhaps similar cost to determining it this causes a deadlock?) and so throwing an error in detected races sounds good to me.

Reply:
> cheaply prevent the use of the channel from any task that is not (transitively) part of the `@sync` group

I guess we can lower `@sync $body` to

```julia
task_local_storage(SYNC_KEY, Channel(Inf)) do
    $body
end
```

and `@spawn $body` to

```julia
let ch = get(task_local_storage(), SYNC_KEY, nothing)
    function wrapper()
        task_local_storage(SYNC_KEY, ch)
        $body
    end
    t = Task(wrapper)
    t.sticky = false
    if ch !== nothing
        put!(ch, t)
    end
    schedule(t)
end
```

?

Not sure if you'd consider `task_local_storage` reasonably cheap, though :)

But maybe more importantly, it's a breaking change since I think it's reasonable to write something like

```julia
@sync begin
    Threads.foreach(xs) do x
        y = cpu_intensive_work(x)
        if rare_condition(y)
            @spawn io_intensive_work(y)
        end
    end
end
```

Reply:
I tweaked `sync_end` in 19a946f7b1968a2b5edf048706c6ee88c02b2725. Now it won't wait for the racy tasks and it always throws when a racy task is found.